src/deploy/functions/functionsDeployHelper.ts

```diff
@@ -127,6 +127,84 @@ export function getEndpointFilters(options: { only?: string }): EndpointFilter[]
 /**
  * Generate label for a function.
  */
-export function getFunctionLabel(fn: backend.TargetIds): string {
-  return `${fn.id}(${fn.region})`;
+export function getFunctionLabel(fn: backend.TargetIds & { codebase?: string }): string {
+  let id = `${fn.id}(${fn.region})`;
+  if (fn.codebase && fn.codebase !== DEFAULT_CODEBASE) {
+    id = `[${fn.codebase}]${id}`;
+  }
+  return id;
+}
+
+/**
+ * Returns list of codebases specified in firebase.json filtered by --only filters if present.
+ */
+export function targetCodebases(config: ValidatedConfig, filters?: EndpointFilter[]): string[] {
+  const codebasesFromConfig = new Set(Object.values(config).map((c) => c.codebase));
```

Comment:
I notice that your using quite a bit of Set's here, I've received feedback from @inlined to try to stay away from them since they're not really idiomatic of js. I see here why they're useful, so I'm a bit conflicted